client/resourcemanager_client.go

```diff
@@ -0,0 +1,282 @@
+		err = errors.WithStack(err)
+		if err != nil {
+			return nil, err
+		}
+		tokenBuckets = req.TokenBuckets
+		return
+	case <-req.requestCtx.Done():
+		return nil, errors.WithStack(req.requestCtx.Err())
+	case <-req.clientCtx.Done():
+		return nil, errors.WithStack(req.clientCtx.Err())
+	}
+}
+
+type tokenBatchController struct {
+	tokenRequestCh chan *tokenRequest
+}
+
+func newTokenBatchController(tokenRequestCh chan *tokenRequest) *tokenBatchController {
+	return &tokenBatchController{
+		tokenRequestCh: tokenRequestCh,
+	}
+}
+
+type tokenDispatcher struct {
+	dispatcherCancel     context.CancelFunc
+	tokenBatchController *tokenBatchController
+}
+
+type resourceManagerConnectionContext struct {
+	stream rmpb.ResourceManager_AcquireTokenBucketsClient
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (c *client) createTokenispatcher() {
+	dispatcherCtx, dispatcherCancel := context.WithCancel(c.ctx)
+	dispatcher := &tokenDispatcher{
+		dispatcherCancel: dispatcherCancel,
+		tokenBatchController: newTokenBatchController(
+			make(chan *tokenRequest, 1)),
+	}
+	go c.handleResouceTokenDispatcher(dispatcherCtx, dispatcher.tokenBatchController)
+	c.tokenDispatcher = dispatcher
+}
+
+func (c *client) handleResouceTokenDispatcher(dispatcherCtx context.Context, tbc *tokenBatchController) {
+	var connection resourceManagerConnectionContext
+	if err := c.tryResourceManagerConnect(dispatcherCtx, &connection); err != nil {
+		log.Warn("get stream error", zap.Error(err))
+	}
+
+	for {
+		var firstRequest *tokenRequest
+		select {
+		case <-dispatcherCtx.Done():
+			return
+		case firstRequest = <-tbc.tokenRequestCh:
+		}
+		stream, streamCtx, cancel := connection.stream, connection.ctx, connection.cancel
+		if stream == nil {
+			c.tryResourceManagerConnect(dispatcherCtx, &connection)
+			c.finishTokenRequest(firstRequest, nil, errors.Errorf("no stream"))
+			continue
+		}
+		select {
+		case <-streamCtx.Done():
+			log.Info("[resource_manager] resource manager stream is canceled")
+			cancel()
+			stream = nil
+			continue
+		default:
+		}
+		err := c.processTokenRequests(stream, firstRequest)
+		if err != nil {
+			log.Info("processTokenRequests error", zap.Error(err))
+			cancel()
+			connection.stream = nil
+		}
+	}
+}
+
+func (c *client) processTokenRequests(stream rmpb.ResourceManager_AcquireTokenBucketsClient, t *tokenRequest) error {
+	req := t.Requeset
+	if err := stream.Send(req); err != nil {
+		err = errors.WithStack(err)
+		c.finishTokenRequest(t, nil, err)
+		return err
+	}
+	resp, err := stream.Recv()
+	if err != nil {
+		err = errors.WithStack(err)
+		c.finishTokenRequest(t, nil, err)
+		return err
+	}
+	if resp.GetError() != nil {
+		return errors.Errorf("[resource_manager]" + resp.GetError().Message)
+	}
+	tokenBuckets := resp.GetResponses()
+	c.finishTokenRequest(t, tokenBuckets, nil)
+	return nil
+}
+
+func (c *client) finishTokenRequest(t *tokenRequest, tokenBuckets []*rmpb.TokenBucketResponse, err error) {
```

Comment:
It seems that this function is too simple to exist. Maybe we can just remove it to avoid some data copying.